proc_macros/src/lib.rs

```diff
@@ -165,3 +203,599 @@ pub fn rule_tests(input: TokenStream) -> TokenStream {
+                })
+            }
+        }
+    }
+}
+
+fn get_rule_instance_struct_definition(
+    rule_struct_name: &Ident,
+    rule_instance_struct_name: &Ident,
+    rule_instance_state_fields: &[&RuleStateFieldSpec],
+) -> proc_macro2::TokenStream {
+    let state_field_names = rule_instance_state_fields.iter().map(|field| &field.name);
+    let state_field_types = rule_instance_state_fields.iter().map(|field| &field.type_);
+    quote! {
+        struct #rule_instance_struct_name {
+            rule: std::sync::Arc<#rule_struct_name>,
+            listener_queries: Vec<crate::rule::RuleListenerQuery>,
+            #(#state_field_names: #state_field_types),*
+        }
+    }
+}
+
+fn get_rule_instance_rule_instance_impl(
+    rule_instance_struct_name: &Ident,
+    rule_instance_per_file_struct_name: &Ident,
+    rule_instance_per_file_state_fields: &[&RuleStateFieldSpec],
+) -> proc_macro2::TokenStream {
+    let rule_instance_per_file_state_field_names = rule_instance_per_file_state_fields
+        .iter()
+        .map(|field| &field.name);
+    let rule_instance_per_file_state_field_initializers = rule_instance_per_file_state_fields
+        .iter()
+        .map(|field| match field.initializer.as_ref() {
+            Some(initializer) => quote!(#initializer),
+            None => quote!(Default::default()),
+        });
+    quote! {
+        impl crate::rule::RuleInstance for #rule_instance_struct_name {
+            fn instantiate_per_file(
+                self: std::sync::Arc<Self>,
+                _file_run_info: &crate::rule::FileRunInfo,
+            ) -> Arc<dyn crate::rule::RuleInstancePerFile> {
+                std::sync::Arc::new(#rule_instance_per_file_struct_name {
+                    rule_instance: self,
+                    #(#rule_instance_per_file_state_field_names: #rule_instance_per_file_state_field_initializers),*
+                })
+            }
+
+            fn rule(&self) -> Arc<dyn crate::rule::Rule> {
+                self.rule.clone()
+            }
+
+            fn listener_queries(&self) -> &[crate::rule::RuleListenerQuery] {
+                &self.listener_queries
+            }
+        }
+    }
+}
+
+fn get_rule_instance_per_file_struct_definition(
+    rule_instance_struct_name: &Ident,
+    rule_instance_per_file_struct_name: &Ident,
+    rule_instance_per_file_state_fields: &[&RuleStateFieldSpec],
+) -> proc_macro2::TokenStream {
+    let state_field_names = rule_instance_per_file_state_fields
+        .iter()
+        .map(|field| &field.name);
+    let state_field_types = rule_instance_per_file_state_fields
+        .iter()
+        .map(|field| &field.type_);
+    quote! {
+        struct #rule_instance_per_file_struct_name {
+            rule_instance: std::sync::Arc<#rule_instance_struct_name>,
+            #(#state_field_names: #state_field_types),*
+        }
+    }
+}
+
+struct SelfAccessRewriter<'a> {
```

Comment:
`syn` has a "mutable visitor" pattern that I used to do the "magic `self.whatever` re-writings"